README.md
# MiniSearch

[![build](https://github.com/micpst/minisearch/actions/workflows/build.yml/badge.svg)](https://github.com/micpst/minisearch/actions/workflows/build.yml)

Restful, in-memory, full-text search engine written in Go.

## ✅ Features

- [x] Full-text indexing of multiple fields in a document
- [x] Boolean queries with AND, OR operators between subqueries
- [x] Exact phrase search
- [x] Document ranking based on BM25
- [x] Vector similarity search for semantic search
- [x] Stemming-based query expansion for many languages
- [x] Document deletion and updating with index garbage collection

## 🛠️ Installation

### Download binary
To download and run minisearch from a precompiled binary:
1. Download a precompiled version of minisearch from GitHub.
2. Run the server binary:
```bash
$ ./server
```

### Run with Docker
To run minisearch with Docker, use the **minisearch** Docker image:
```bash
$ docker run -d --name minisearch -p 3000:3000 micpst/minisearch:latest
```

### Build from source
To build and run minisearch from the source code:
1. Requirements: **go & make**
2. Install dependencies:
```bash
$ make setup
```
3. Build:
```bash
$ make build
```
4. Run the server binary:
```bash
$ ./bin/server
```

## 📘 Usage
### Add documents
Create a new document and add it to the index.
```bash
$ curl -X POST localhost:3000/api/v1/documents \
    -H 'Content-Type: application/json' \
    -d '{ 
      "title": "The Silicon Brain", 
      "url": "https://micpst.com/posts/silicon-brain", 
      "abstract": "The human brain is often described as complex..." 
    }'
```

### Upload document dumps
Fill the index with a large number of documents at once by uploading a document dumps.
```bash
$ curl -X POST localhost:3000/api/v1/upload \
    -H 'Content-Type: multipart/form-data' \
    -F 'file[]=@/path/to/dataset1.xml.gz' \
    -F 'file[]=@/path/to/dataset2.xml.gz'
```
The dump should have the following structure:
```xml
<docs>
  <doc>
    <title>...</title>
    <url>...</url>
    <abstract>...</abstract>
  </doc>
  <doc>
    <title>...</title>
    <url>...</url>
    <abstract>...</abstract>
  </doc>
</docs>
```

### Update the document
Update the existing document and re-index it with the new fields.
```bash
$ curl -X PUT localhost:3000/api/v1/documents/<id> \
    -H 'Content-Type: application/json' \
    -d '{ 
      "title": "The Silicon Brain", 
      "url": "https://micpst.com/posts/silicon-brain", 
      "abstract": "The human brain is often described as complex..." 
    }'
```

### Remove the document
Permanently delete the document and remove it from the index.
```bash
$ curl -X DELETE localhost:3000/api/v1/documents/<id>
```

### Search the index

#### Search properties
The `properties` parameter defines in which property to run our query.
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Brain",
      "properties": ["title"]
    }'
```
We are now searching for all the documents that contain the word `Brain` in the `title` property.

We can also search through nested properties:
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Mic",
      "properties": ["author.name"],
    }'
```
By default, MiniSearch searches in all searchable properties.

#### Exact match
The `exact` property finds all the document with an exact match of the `query` property.
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Brain",
      "properties": ["title"],
      "exact": true
    }'
```
We are now searching for all the documents that contain `exactly` the word `Brain` in the `title` property.

> Without the `exact` property, for example, the term `Brain-busting` would be returned as well, as it contains the word `Brain`.

#### Typo tolerance
The `tolerance` property allows specifying the maximum distance (following the Levenshtein algorithm) between the query and the searchable property.
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Brin",
      "properties": ["title"],
      "tolerance": 1
    }'
```
We are searching for all the documents that contain a term with an edit distance of `1` (e.g. `Brain`) in the `title` property.

> `tolerance` doesn't work together with the `exact` parameter. `exact` will have priority.

#### Pagination
The `offset` and `limit` properties allow paginating the results.
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Brain",
      "properties": ["title"],
      "offset": 10,
      "limit": 5
    }'
```
By default, MiniSearch limits the search results to 10, without any offset.

#### BM25 ranking
MiniSearch uses the BM25 algorithm to calculate the relevance of a document when searching.

You can edit the BM25 parameters by using the `relevance` property in the `search` configuration object.
```bash
$ curl -X POST localhost:3000/api/v1/search \
    -H 'Content-Type: application/json' \
    -d '{
      "query": "Brain",
      "properties": ["title"],
      "relevance": {
        // Term frequency saturation parameter.
        // Default value: 1.2
        // Recommended value: between 1.2 and 2
        "k": 1.2,
        
        // Length normalization parameter.
        // Default value: 0.75
        // Recommended value: > 0.75
        "b": 0.75,
        
        // Frequency normalization lower bound.
        // Default value: 0.5
        // Recommended value: between 0.5 and 1
        "d": 0.5
      }
    }'
```

## 📄 License
All my code is MIT licensed. Libraries follow their respective licenses.


CHANGELOG.md
# Changelog

## [1.2.0] 2023-04-28

### Added:
- Search results ranking based on BM25
- Stemming-based query expansion for many languages
- Vector similarity search for semantic search

### Changed:
- Change search API endpoint HTTP method to POST
- Move search params from query string to request body

## [1.1.0] 2023-02-26

### Added:
- Search results ranking based on TF-IDF
- Results pagination

### Changed:
- Rename project to `minisearch`

## [1.0.1] 2023-02-24

### Changed:
- Bump go version to 1.20
- Update dependencies
- Improve overall performance

### Removed:
- Remove `github.com/cornelk/hashmap` dependency

## [1.0.0] 2023-02-13

### Added:
- Full-text indexing of multiple fields in a document
- Boolean queries with AND, OR operators between subqueries
- Document deletion and updating with index garbage collection 


api/handlers.go
package api

import (
	"compress/gzip"
	"encoding/xml"
	"mime/multipart"
	"net/http"
	"strings"
	"time"
	"unsafe"

	"github.com/gin-gonic/gin"
	"github.com/micpst/minisearch/pkg/store"
	"github.com/micpst/minisearch/pkg/tokenizer"
)

type DocumentResponse struct {
	Id       string `json:"id"`
	Title    string `json:"title"`
	Url      string `json:"url"`
	Abstract string `json:"abstract"`
}

type SearchDocument struct {
	Id    string   `json:"id"`
	Data  Document `json:"data"`
	Score float64  `json:"score"`
}

type SearchDocumentResponse struct {
	Count   int              `json:"count"`
	Hits    []SearchDocument `json:"hits"`
	Elapsed int64            `json:"elapsed"`
}

type UploadDocumentsResponse struct {
	Total   int `json:"total"`
	Success int `json:"success"`
	Fail    int `json:"fail"`
}

type UploadDocumentsFileDump struct {
	Documents []Document `xml:"doc"`
}

func (s *Server) uploadDocuments(c *gin.Context) {
	form, err := c.MultipartForm()
	if err != nil {
		c.Status(http.StatusBadRequest)
		return
	}

	total := 0
	failed := 0
	files := form.File["file[]"]

	for _, file := range files {
		dump, err := loadDocumentsFromFile(file)
		if err != nil {
			c.Status(http.StatusBadRequest)
			return
		}

		params := store.InsertBatchParams[Document]{
			Documents: dump.Documents,
			BatchSize: 10000,
			Language:  tokenizer.Language(strings.ToLower(c.Query("lang"))),
		}
		errs := s.db.InsertBatch(&params)

		total += len(dump.Documents)
		failed += len(errs)
	}

	c.JSON(http.StatusOK, UploadDocumentsResponse{
		Total:   total,
		Success: total - failed,
		Fail:    failed,
	})
}

func (s *Server) createDocument(c *gin.Context) {
	body := Document{}
	if err := c.BindJSON(&body); err != nil {
		return
	}

	params := store.InsertParams[Document]{
		Document: body,
		Language: tokenizer.Language(strings.ToLower(c.Query("lang"))),
	}

	doc, err := s.db.Insert(&params)
	if err != nil {
		c.Status(http.StatusInternalServerError)
		return
	}

	c.JSON(http.StatusCreated, documentFromRecord(doc))
}

func (s *Server) updateDocument(c *gin.Context) {
	body := Document{}
	if err := c.BindJSON(&body); err != nil {
		return
	}

	params := store.UpdateParams[Document]{
		Id:       c.Param("id"),
		Document: body,
		Language: tokenizer.Language(strings.ToLower(c.Query("lang"))),
	}

	doc, err := s.db.Update(&params)
	if err != nil {
		c.Status(http.StatusNotFound)
		return
	}

	c.JSON(http.StatusOK, documentFromRecord(doc))
}

func (s *Server) deleteDocument(c *gin.Context) {
	params := store.DeleteParams[Document]{
		Id:       c.Param("id"),
		Language: tokenizer.Language(strings.ToLower(c.Query("lang"))),
	}

	if err := s.db.Delete(&params); err != nil {
		c.Status(http.StatusNotFound)
		return
	}

	c.Status(http.StatusOK)
}

func (s *Server) searchDocuments(c *gin.Context) {
	params := store.SearchParams{
		Properties: []string{},
		BoolMode:   store.AND,
		Limit:      10,
		Relevance: store.BM25Params{
			K: 1.2,
			B: 0.75,
			D: 0.5,
		},
	}
	if err := c.BindJSON(&params); err != nil {
		return
	}

	start := time.Now()
	result, err := s.db.Search(&params)
	elapsed := time.Since(start)

	if err != nil {
		c.Status(http.StatusBadRequest)
		return
	}

	c.JSON(http.StatusOK, SearchDocumentResponse{
		Count:   result.Count,
		Hits:    *(*[]SearchDocument)(unsafe.Pointer(&result.Hits)),
		Elapsed: elapsed.Microseconds(),
	})
}

func documentFromRecord(d store.Record[Document]) DocumentResponse {
	return DocumentResponse{
		Id:       d.Id,
		Title:    d.Data.Title,
		Url:      d.Data.Url,
		Abstract: d.Data.Abstract,
	}
}

func loadDocumentsFromFile(file *multipart.FileHeader) (UploadDocumentsFileDump, error) {
	f, err := file.Open()
	defer func(f multipart.File) {
		_ = f.Close()
	}(f)
	if err != nil {
		return UploadDocumentsFileDump{}, err
	}

	gz, err := gzip.NewReader(f)
	defer func(gz *gzip.Reader) {
		_ = gz.Close()
	}(gz)
	if err != nil {
		return UploadDocumentsFileDump{}, err
	}

	d := xml.NewDecoder(gz)
	dump := UploadDocumentsFileDump{}
	if err := d.Decode(&dump); err != nil {
		return UploadDocumentsFileDump{}, err
	}

	return dump, nil
}
